nucleus/common/common-util/src/main/java/com/sun/enterprise/universal/process/ProcessManager.java

```diff
@@ -222,51 +211,66 @@ private void waitAwhile(Process process) throws InterruptedException, ProcessMan
     }
 
 
-    private void doWaitForReaderThreads() {
-        if (waitForReaderThreads) {
-            // wait for stdin and stderr to finish up
-            for (Thread t : threads) {
-                try {
-                    t.join();
-                } catch (InterruptedException ex) {
-                    // nothing to do
-                }
-            }
-        }
-    }
-
-
-    static class ReaderThread implements Runnable {
+    static class ReaderThread extends Thread {
         private final BufferedReader reader;
-        private final StringBuffer sb;
+        private final StringBuilder sb;
         private final boolean echo;
+        private boolean stop;
```

Comment:
Hmm, non-volatile flag enough here?

Reply:
I believe it is not required here. If it would be several nanoseconds later it doesn't matter. Perhaps we should make some experiments if it is better to use volatile and slow down some context switching or not to use volatile and slow down the "shutdown" (it can make few iterations more).

Reply:
The `stop` flag set to true in the thread running `execute` (in `finish` method), but check in another (reader) thread. I think, visibility problem is possible here. We can get an infinite wait loop.

Reply:
Volatile is not a solution for thread safety, just an instruction to avoid optimization. The value change has just one direction, from false to true, and then with volatile it must be checked in memory before every use, without volatile it can be used from cpu cache between syncing the value state in memory, which is done by the cpu.
Or am I wrong and cpu can use the old value forever?
Can you prove that behavior on some cpu?

Reply:
Volatile write give us StoreLoad barrier to flush CPU write buffers. Without it cpu may use value from its local wtite buffer potentially indefinitely, i.e., second thread, running on another core/cpu may never see new value.

I have no cpu other than x86 (amd and intel) ;-(

Reply:
Hmm, I know about related issues if we would have two depending values, but in this case it seems contraproductive to me. But we don't need to hunt nanos here, so for simplicity I can just add it.

I am just curious if we are not overengineering that. Just to remember ie if arms would have such issue... and how to reproduce it.

Reply:
In case of two variables even volatile may not be sufficient ;-)

Reply:
As a microoptimization we can use AtomicBoolen lazySet/get here

Reply:
I had to read more ... now I am thinking about some tests outside GF related to this problematic :-)
https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/invoke/VarHandle.html
https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/atomic/AtomicBoolean.html

I used acquire-release, finally.
I tried to google where we could reproduce some issues like never ending loop, somebody mentioned Apple and ARM cpus.

Reply:
Great! I think, release/acquire enough here.